explorer/interpreter/type_checker.cpp

```diff
@@ -1635,6 +1688,42 @@ auto TypeChecker::Substitute(const Bindings& bindings,
   }
 }
 
+auto TypeChecker::RefineWitness(Nonnull<const Witness*> witness,
+                                Nonnull<const Value*> type,
+                                Nonnull<const Value*> constraint) const
+    -> Nonnull<const Witness*> {
+  if (!top_level_impl_scope_) {
+    return witness;
+  }
+
+  // See if this is already resolved as some number of layers of
+  // ConstraintImplWitness applied to an ImplWitness.
+  Nonnull<const Witness*> inner_witness = witness;
+  while (auto* inner_constraint_impl_witness =
+             dyn_cast<ConstraintImplWitness>(inner_witness)) {
+    inner_witness = inner_constraint_impl_witness->constraint_witness();
+  }
+  if (isa<ImplWitness>(inner_witness)) {
+    return witness;
+  }
+
+  // No source location; diagnostics will be discarded.
+  SourceLocation source_loc("", 0);
```

Comment:
Kind of wondering if we should add a SourceLocation::Unknown or similar, I think there are a few similar examples (but I also generally tried to attach _something_ when possible, at least a file-like location)

Reply:
Yeah, I think it's worth adding a named factory function to make it easier to audit the places where we do this. (Ideally there should not be many of them!) Will send a separate PR.